server/src/main/java/org/elasticsearch/plugins/IndexStorePlugin.java

```diff
@@ -82,4 +86,41 @@
     default Map<String, RecoveryStateFactory> getRecoveryStateFactories() {
         return Collections.emptyMap();
     }
+
+    /**
+     * {@link IndexFoldersDeletionListener} are invoked before the folders of a shard or an index are deleted from disk.
+     */
+    interface IndexFoldersDeletionListener {
+        /**
+         * Invoked before the folders of an index are deleted from disk. The list of folders contains {@link Path}s that may or may not
+         * exist on disk. Shard locks are expected to be acquired at the time this method is invoked.
+         *
+         * @param index         the {@link Index} of the index whose folders are going to be deleted
+         * @param indexSettings settings for the index whose folders are going to be deleted
+         * @param indexPaths    the paths of the folders that are going to be deleted
+         */
+        default void beforeIndexFoldersDeleted(Index index, IndexSettings indexSettings, List<Path> indexPaths) {
```

Comment:
I think we don't need to supply a no-op default here, we can reasonably require implementations to implement both methods.

Reply:
I pushed 5925e5afa90c278be9bb2388d1ccfe2670609747